lib/pure/asyncdispatch.nim

```diff
@@ -640,45 +635,15 @@ when defined(windows) or defined(nimdoc):
 
     let ret = WSARecv(socket.SocketHandle, addr dataBuf, 1, addr bytesReceived,
                       addr flagsio, cast[POVERLAPPED](ol), nil)
-    if ret == -1:
+    # Processing of "immediate completion" operations was removed, because under
+    # heavy load it can lead to deep recursion and stack overflow.
+    # This change also dont affect speed.
+    if ret != 0:
       let err = osLastError()
       if err.int32 != ERROR_IO_PENDING:
-        if dataBuf.buf != nil:
-          dealloc dataBuf.buf
-          dataBuf.buf = nil
+        dealloc dataBuf.buf
         GC_unref(ol)
-        if flags.isDisconnectionError(err):
```

Comment:
Please remember your answer in IRC

> **cheatfate_:** dom96, do you really sure of the correctness of your isDisconnectionError() from net.nim?
> **dom96:** yes

Now i will tell you about results of 2 tests i have made:
Test 1
We have host machine on windows with running slightly modified version of asynchttpserver.nim to show return value of **WSARecv()** and corresponding **osLastError()** and arguments of callback function **bytesCount** and **errcode**. Also added timeout so first **WSARecv()** after connection will be called only when timeout expires.

We have client machine on Linux, with telnet <host> 5555.

We connecting with telnet from linux to windows ip and port... windows host is start timeout waiting, after that before timeout expired i have disabled linux interface. So there no packets can come from linux to windows.

Result: WSARecv() returned **-1**, osLastErrorCode() returned 997 (ERROR_IO_PENDING). callback was not called...

Second test was made like first except i have disabled windows interface while asynchttpserver waiting to make first WSARecv().

Result: WSARecv() returned **-1**, osLastErrorCode() returned 997 (ERROR_IO_PENDING). callback was not called...

There also one more test needs to be done to be completely sure: Sending RST tcp packet to host after connection have been made.

Currently i can say that bunch of error codes used in isDisconnectionError() not working with WSARecv() in OVERLAPPED IO mode.